Using **Flight Operations** dataset to perform Complete Data Analysis

In [1]:
import pandas as pd

#Load the dataset

In [2]:
df = pd.read_csv(r'/content/drive/MyDrive/Colab Notebooks/IDRA/Day10_Flight_Operations_Dataset.csv')

#Inspecting dataset

In [3]:
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5


In [4]:
df.tail()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
175,FL0176,2026-04-10,IndiGo,Ahmedabad,Bengaluru,Airbus A320,Economy,48,189,4298,0.0,On Time,Storm,Mobile App,11,Vegetarian,5
176,FL0177,2026-02-17,Vistara,Chennai,Mumbai,Airbus A320,Business,124,180,17306,20.0,Delayed,Cloudy,Mobile App,8,No Meal,5
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3
178,FL0179,2026-05-08,Vistara,Bengaluru,Kochi,Airbus A320,Economy,134,189,5413,20.0,Delayed,Rain,Online Travel Portal,9,No Meal,2
179,FL0180,2026-01-22,Akasa Air,Delhi,Mumbai,Airbus A321,Economy,144,210,3767,75.0,Delayed,Fog,Online Travel Portal,17,No Meal,5


In [5]:
df.sample(5)

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
58,FL0059,2026-01-10,Air India Express,Delhi,Srinagar,Airbus A319,Economy,159,220,3708,20.0,Delayed,Storm,Airline Website,11,No Meal,2
87,FL0088,2026-02-08,Vistara,Delhi,Srinagar,Airbus A321,Economy,75,210,4156,0.0,On Time,Cloudy,Travel Agency,20,Vegetarian,3
44,FL0045,2026-06-05,SpiceJet,Delhi,Srinagar,Airbus A320,Economy,81,186,3975,35.0,Delayed,Clear,Travel Agency,14,Non-Vegetarian,3
70,FL0071,2026-03-31,SpiceJet,Lucknow,Delhi,Boeing 737,Premium Economy,101,180,10380,5.0,On Time,Cloudy,Online Travel Portal,16,Vegetarian,2
53,FL0054,2026-04-11,SpiceJet,Mumbai,Bengaluru,Airbus A320,Economy,49,180,4738,20.0,Delayed,Rain,Mobile App,14,No Meal,3


In [6]:
df.shape

(180, 17)

In [7]:
df.columns

Index(['Flight_ID', 'Flight_Date', 'Airline', 'Origin', 'Destination',
       'Aircraft', 'Travel_Class', 'Passengers', 'Seat_Capacity',
       'Average_Ticket_Price', 'Delay_Minutes', 'Flight_Status', 'Weather',
       'Booking_Channel', 'Avg_Baggage_Kg', 'Meal_Preference',
       'Passenger_Satisfaction'],
      dtype='object')

In [8]:
df.index

RangeIndex(start=0, stop=180, step=1)

In [9]:
df.dtypes

,0
Flight_ID,object
Flight_Date,object
Airline,object
Origin,object
Destination,object
Aircraft,object
Travel_Class,object
Passengers,int64
Seat_Capacity,int64
Average_Ticket_Price,int64


In [10]:
df.isnull().sum()

,0
Flight_ID,0
Flight_Date,0
Airline,0
Origin,0
Destination,0
Aircraft,0
Travel_Class,0
Passengers,0
Seat_Capacity,0
Average_Ticket_Price,0


In [11]:
df.describe()

,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Avg_Baggage_Kg,Passenger_Satisfaction
count,180.000000,180.000000,180.000000,173.000000,180.000000,180.000000
mean,122.772222,197.238889,6111.677778,19.335260,15.150000,3.494444
std,45.635651,15.408176,3615.220040,29.029712,5.705506,1.169908
min,45.000000,180.000000,1949.000000,0.000000,5.000000,2.000000
25%,78.750000,186.000000,3815.500000,0.000000,10.000000,2.000000
50%,125.000000,189.000000,5380.000000,5.000000,15.500000,3.000000
75%,162.250000,210.000000,6790.000000,20.000000,20.000000,5.000000
max,210.000000,220.000000,27832.000000,110.000000,25.000000,5.000000


In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 180 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               180 non-null    object 
 1   Flight_Date             180 non-null    object 
 2   Airline                 180 non-null    object 
 3   Origin                  180 non-null    object 
 4   Destination             180 non-null    object 
 5   Aircraft                180 non-null    object 
 6   Travel_Class            180 non-null    object 
 7   Passengers              180 non-null    int64  
 8   Seat_Capacity           180 non-null    int64  
 9   Average_Ticket_Price    180 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           180 non-null    object 
 12  Weather                 180 non-null    object 
 13  Booking_Channel         180 non-null    object 
 14  Avg_Baggage_Kg          180 non-null    in

We can observe that there is
- 180 rows
- 17 columns
- Flight_Date is an object, not in datetime datatype
- 7 null elements in Delay_Minutes column

In [13]:
df.drop_duplicates(inplace=True)
df.dropna(inplace=True)
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 173 entries, 0 to 179
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Flight_ID               173 non-null    object 
 1   Flight_Date             173 non-null    object 
 2   Airline                 173 non-null    object 
 3   Origin                  173 non-null    object 
 4   Destination             173 non-null    object 
 5   Aircraft                173 non-null    object 
 6   Travel_Class            173 non-null    object 
 7   Passengers              173 non-null    int64  
 8   Seat_Capacity           173 non-null    int64  
 9   Average_Ticket_Price    173 non-null    int64  
 10  Delay_Minutes           173 non-null    float64
 11  Flight_Status           173 non-null    object 
 12  Weather                 173 non-null    object 
 13  Booking_Channel         173 non-null    object 
 14  Avg_Baggage_Kg          173 non-null    int64  

In [14]:
df['Flight_Date'] = pd.to_datetime(df['Flight_Date'])
df.dtypes

,0
Flight_ID,object
Flight_Date,datetime64[ns]
Airline,object
Origin,object
Destination,object
Aircraft,object
Travel_Class,object
Passengers,int64
Seat_Capacity,int64
Average_Ticket_Price,int64


After cleaning the dataset, it is clear that
- Flight_Date is now a datetime datatype
- 180 rows is changed to 173 rows

#Data Selection

In [15]:
df['Airline']

,Airline
0,SpiceJet
1,Akasa Air
2,Air India
3,Akasa Air
4,Air India
...,...
175,IndiGo
176,Vistara
177,IndiGo
178,Vistara


In [16]:
df[['Flight_ID', 'Airline', 'Aircraft']]

,Flight_ID,Airline,Aircraft
0,FL0001,SpiceJet,Airbus A319
1,FL0002,Akasa Air,Boeing 737
2,FL0003,Air India,Boeing 737
3,FL0004,Akasa Air,Airbus A319
4,FL0005,Air India,Airbus A319
...,...,...,...
175,FL0176,IndiGo,Airbus A320
176,FL0177,Vistara,Airbus A320
177,FL0178,IndiGo,Airbus A319
178,FL0179,Vistara,Airbus A320


In [17]:
df.loc[df['Flight_ID'].isin(['FL0001','FL0005']), ['Flight_ID', 'Airline', 'Aircraft']]

,Flight_ID,Airline,Aircraft
0,FL0001,SpiceJet,Airbus A319
4,FL0005,Air India,Airbus A319


#Filtering

In [18]:
df[df['Origin']=='Delhi']

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5
6,FL0007,2026-06-17,Air India Express,Delhi,Srinagar,Airbus A319,Premium Economy,178,210,9430,0.0,On Time,Storm,Online Travel Portal,18,Vegetarian,4
7,FL0008,2026-04-28,Air India,Delhi,Chennai,Airbus A319,Business,157,220,14173,0.0,On Time,Fog,Airline Website,25,Vegan,5
8,FL0009,2026-01-25,Akasa Air,Delhi,Chennai,Airbus A320,Economy,82,210,5893,75.0,Delayed,Cloudy,Mobile App,8,Vegetarian,4
16,FL0017,2026-05-24,SpiceJet,Delhi,Chennai,Airbus A321,Economy,73,220,4256,20.0,Delayed,Rain,Travel Agency,22,Non-Vegetarian,2
22,FL0023,2026-02-02,Akasa Air,Delhi,Mumbai,Airbus A319,Economy,52,210,4620,12.0,On Time,Fog,Mobile App,9,Non-Vegetarian,2
29,FL0030,2026-04-10,SpiceJet,Delhi,Srinagar,Airbus A321,Economy,200,210,2716,75.0,Delayed,Storm,Airline Website,10,Vegan,3
30,FL0031,2026-01-08,Vistara,Delhi,Srinagar,Airbus A321,Premium Economy,209,180,9962,50.0,Delayed,Cloudy,Mobile App,9,Vegetarian,4
33,FL0034,2026-06-18,Air India Express,Delhi,Chennai,Boeing 737,Premium Economy,125,186,9278,0.0,On Time,Rain,Online Travel Portal,23,No Meal,3
34,FL0035,2026-04-24,Vistara,Delhi,Srinagar,Boeing 737,Premium Economy,77,210,11441,0.0,On Time,Fog,Travel Agency,9,Vegan,3


In [19]:
df[df['Delay_Minutes']>100]

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
26,FL0027,2026-02-10,SpiceJet,Bengaluru,Hyderabad,Airbus A319,Economy,169,180,3973,110.0,Delayed,Clear,Mobile App,7,Vegan,4
50,FL0051,2026-05-03,Air India,Delhi,Mumbai,Airbus A321,Economy,135,210,5129,110.0,Delayed,Rain,Online Travel Portal,17,Vegetarian,4
57,FL0058,2026-04-14,IndiGo,Hyderabad,Kolkata,Boeing 737,Premium Economy,151,210,6834,110.0,Delayed,Cloudy,Travel Agency,16,Vegetarian,4
82,FL0083,2026-04-18,Air India Express,Lucknow,Delhi,Airbus A319,Economy,71,180,2484,110.0,Delayed,Cloudy,Online Travel Portal,22,Vegetarian,4
128,FL0129,2026-01-04,SpiceJet,Lucknow,Delhi,Boeing 737,Premium Economy,67,210,5657,110.0,Delayed,Cloudy,Online Travel Portal,21,Non-Vegetarian,2
146,FL0147,2026-02-18,IndiGo,Mumbai,Kolkata,Airbus A319,Premium Economy,115,210,9106,110.0,Delayed,Rain,Online Travel Portal,19,Vegetarian,5
161,FL0162,2026-01-18,Air India,Hyderabad,Kolkata,Airbus A321,Economy,128,189,5263,110.0,Delayed,Rain,Mobile App,6,Non-Vegetarian,3
169,FL0170,2026-06-25,IndiGo,Mumbai,Kolkata,Airbus A319,Economy,130,180,3938,110.0,Delayed,Rain,Online Travel Portal,17,No Meal,4
177,FL0178,2026-04-30,IndiGo,Delhi,Mumbai,Airbus A319,Economy,112,186,4490,110.0,Delayed,Fog,Online Travel Portal,16,Vegetarian,3


In [20]:
df[df['Flight_Date']=='2026-04-03']

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction
80,FL0081,2026-04-03,SpiceJet,Delhi,Mumbai,Airbus A320,Economy,199,180,2378,50.0,Delayed,Rain,Online Travel Portal,13,No Meal,4


#Sorting

In [21]:
print("Highly delayed flights")
delayed = df['Delay_Minutes'].sort_values(ascending=False)
delayed.head(10)

Highly delayed flights


,Delay_Minutes
26,110.0
177,110.0
82,110.0
57,110.0
50,110.0
128,110.0
146,110.0
169,110.0
161,110.0
21,75.0


In [22]:
print('Flights with highest average ticket price')
df['Average_Ticket_Price'].sort_values(ascending=False).head()

Flights with highest average ticket price


,Average_Ticket_Price
138,27832
52,21499
32,17625
176,17306
108,16390


In [23]:
print('Flights with lowest average ticket price')
df['Average_Ticket_Price'].sort_values().head()

Flights with lowest average ticket price


,Average_Ticket_Price
136,1949
156,2188
90,2251
165,2314
80,2378


#Groupby & aggregation

In [24]:
print("No. of flights wrt Airlines:\n", df.groupby('Airline')['Flight_ID'].count())

No. of flights wrt Airlines:
 Airline
Air India            23
Air India Express    24
Akasa Air            30
IndiGo               32
SpiceJet             29
Vistara              35
Name: Flight_ID, dtype: int64


In [25]:
print("Average delay wrt Airlines:\n", df.groupby('Airline')['Delay_Minutes'].mean())

Average delay wrt Airlines:
 Airline
Air India            21.260870
Air India Express    11.291667
Akasa Air            14.400000
IndiGo               25.031250
SpiceJet             24.137931
Vistara              18.628571
Name: Delay_Minutes, dtype: float64


Observations
- Vistara is the frequently used airline
- IndiGo has the most delays among the airlines

In [26]:
print("Flights wrt their destination:\n", df.groupby('Destination')['Flight_ID'].count())

Flights wrt their destination:
 Destination
Bengaluru    27
Chennai      16
Delhi        49
Hyderabad    13
Kochi         8
Kolkata      20
Mumbai       28
Srinagar     12
Name: Flight_ID, dtype: int64


In [27]:
print("No. of flights with destination Kochi: ", df.groupby('Destination')['Flight_ID'].count()['Kochi'])

No. of flights with destination Kochi:  8


Observations
- to Kochi, 8 flights
- Delhi is the destination of most of the flights

In [28]:
print("Flights (status-wise):\n", df.groupby('Flight_Status')['Flight_ID'].count())

Flights (status-wise):
 Flight_Status
Delayed     59
On Time    114
Name: Flight_ID, dtype: int64


#Data transformation

In [29]:
def occ_rate(row):
  return (row['Passengers']/row['Seat_Capacity'])*100

df['Occupancy_Rate'] = df.apply(occ_rate, axis=1)
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Occupancy_Rate
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,116.666667
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5,33.333333
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,30.454545
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,69.545455
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,98.888889


In [30]:
def avail_seats(row):
  return row['Seat_Capacity'] - row['Passengers']

df['Available_Seats'] = df.apply(avail_seats, axis=1)
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Occupancy_Rate,Available_Seats
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,116.666667,-30
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5,33.333333,140
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,30.454545,153
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,69.545455,67
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,98.888889,2


There are negative values, which could indicate there are people in the waiting list

In [31]:
def delay_status(row):
  x = row['Delay_Minutes']
  if x == 0:
    return 'On Time'
  elif x>0 and x<=10:
    return 'Slight Delay'
  elif x>10 and x<=30:
    return 'Moderate Delay'
  else:
    return 'Severe Delay'

df['Delay_Condition'] = df.apply(delay_status, axis=1)
df.head()

,Flight_ID,Flight_Date,Airline,Origin,Destination,Aircraft,Travel_Class,Passengers,Seat_Capacity,Average_Ticket_Price,Delay_Minutes,Flight_Status,Weather,Booking_Channel,Avg_Baggage_Kg,Meal_Preference,Passenger_Satisfaction,Occupancy_Rate,Available_Seats,Delay_Condition
0,FL0001,2026-03-23,SpiceJet,Mumbai,Bengaluru,Airbus A319,Economy,210,180,6894,50.0,Delayed,Storm,Online Travel Portal,23,No Meal,3,116.666667,-30,Severe Delay
1,FL0002,2026-06-07,Akasa Air,Delhi,Mumbai,Boeing 737,Economy,70,210,4468,0.0,On Time,Fog,Travel Agency,16,Vegan,5,33.333333,140,On Time
2,FL0003,2026-05-12,Air India,Kochi,Delhi,Boeing 737,Economy,67,220,4713,5.0,On Time,Rain,Travel Agency,19,No Meal,2,30.454545,153,Slight Delay
3,FL0004,2026-05-30,Akasa Air,Srinagar,Delhi,Airbus A319,Economy,153,220,3791,35.0,Delayed,Storm,Travel Agency,18,No Meal,2,69.545455,67,Severe Delay
4,FL0005,2026-02-27,Air India,Pune,Delhi,Airbus A319,Premium Economy,178,180,6505,20.0,Delayed,Cloudy,Airline Website,25,No Meal,5,98.888889,2,Moderate Delay


#Airline comparison based on passengers, delays and customer satisfaction

In [32]:
grp = df.groupby('Airline')
summary = grp.agg(
    Num_Flights = ('Flight_ID', 'count'),
    Total_Passengers = ('Passengers', 'sum'),
    Avg_Delay = ('Delay_Minutes', 'mean'),
    Avg_Satisfaction = ('Passenger_Satisfaction', 'mean')
)
print(summary)

                   Num_Flights  Total_Passengers  Avg_Delay  Avg_Satisfaction
Airline                                                                      
Air India                   23              2908  21.260870          3.695652
Air India Express           24              3027  11.291667          3.666667
Akasa Air                   30              3257  14.400000          3.433333
IndiGo                      32              4158  25.031250          3.343750
SpiceJet                    29              3629  24.137931          3.448276
Vistara                     35              4225  18.628571          3.400000


From the summary table,
- most used airline = Vistara
- most delayed airline = IndiGo
- least dealyed airline = Air India Express
- most satisfactory = Air India


#Observation
- most used airline = Vistara
- most delayed airline = IndiGo
- least dealyed airline = Air India Express
- most satisfactory = Air India
- most travelled destination = Delhi